feat: Inverse schema

fritzduchardt opened the pull request:
As a proposal, I have switched around the ytt order of files to include a full schema of the argocd settings in myks. This can be selectively overwritten in the data-schema.ytt.yaml of the client project, the idea being that the client benefits from a simple data-schema.ytt.yaml and sensible defaults.

This change requires `#@overlay/match-child-defaults missing_ok=True` to be added to the client's data-schema.ytt.yaml.

Also, I made an opinionated change to the argo plugin files merging them both into one file with the intention to avoid duplicated methods and aggregated all argo related code in a single file.

Lastly, I prepended the the argo application CR name with the env name since applications would otherwise be likely to overwrite each other.

Zebradil commented:
> As a proposal, I have switched around the ytt order of files to include a full schema of the argocd settings in myks. This can be selectively overwritten in the data-schema.ytt.yaml of the client project, the idea being that the client benefits from a simple data-schema.ytt.yaml and sensible defaults.

If I remember correctly, we were talking about encapsulating the default `data-schema.ytt.yaml` in `myks` and letting users do as they wish (in this case a `envs/data-schema.ytt.yaml` is not required). I'd go this way instead of extracting just the `argocd` part.

This is actually the last thing I'd like to implement before releasing the new version. If you have any suggestions for more breaking changes we'd better implement them now.

> Also, I made an opinionated change to the argo plugin files merging them both into one file with the intention to avoid duplicated methods and aggregated all argo related code in a single file.

Merging the code into a single file makes sense, that's better. However, I'm not sure about new function names.

For example, `app.renderArgoCDApplication` might mean that the function renders the `Application` custom resource of ArgoCD. This is fine in general, but there are no other functions attached to the `Application` structure that render other ArgoCD`s custom resources. Having `app.renderArgoCD` to render all ArgoCD resources related to the current application feels lighter.

In contrast to `app.renderArgoCDApplication`, `env.renderArgoCDEnvironment` doesn't render the `Environment` custom resource. Instead, it renders an `AppProject` and a `Secret`. In this case, having `env.renderArgoCD` seems even more sensible.

Maybe I'm missing something.

> Lastly, I prepended the the argo application CR name with the env name since applications would otherwise be likely to overwrite each other.

I wonder if applications will conflict if they're placed to separate namespaces. I don't really like this change, because it produces more customized internal logic, but let's have it as we both do that at the moment (we in South Pole actually don't need to have this, but my colleague said that he likes this more :-) ).

fritzduchardt commented:
> > As a proposal, I have switched around the ytt order of files to include a full schema of the argocd settings in myks. This can be selectively overwritten in the data-schema.ytt.yaml of the client project, the idea being that the client benefits from a simple data-schema.ytt.yaml and sensible defaults.
> 
> If I remember correctly, we were talking about encapsulating the default `data-schema.ytt.yaml` in `myks` and letting users do as they wish (in this case a `envs/data-schema.ytt.yaml` is not required). I'd go this way instead of extracting just the `argocd` part.

Shouldn't the user be able to add a schema of sorts, in case he wishes to add his own values, e.g. generic settings that affect multiple applications? We use that already in one of our rollout repos.

I have changed the logic to include an entire schema file (the one from assets) and made if flexible to be overwritten with a schema provided by the user. Please have a look. If you like you can make further changes on this PR so get this feature completed.

> 
> This is actually the last thing I'd like to implement before releasing the new version. If you have any suggestions for more breaking changes we'd better implement them now.
> 
> > Also, I made an opinionated change to the argo plugin files merging them both into one file with the intention to avoid duplicated methods and aggregated all argo related code in a single file.
> 
> Merging the code into a single file makes sense, that's better. However, I'm not sure about new function names.
> 
> For example, `app.renderArgoCDApplication` might mean that the function renders the `Application` custom resource of ArgoCD. This is fine in general, but there are no other functions attached to the `Application` structure that render other ArgoCD`s custom resources. Having `app.renderArgoCD` to render all ArgoCD resources related to the current application feels lighter.
> 

Agreed - no need to add that suffix to the method names.

> In contrast to `app.renderArgoCDApplication`, `env.renderArgoCDEnvironment` doesn't render the `Environment` custom resource. Instead, it renders an `AppProject` and a `Secret`. In this case, having `env.renderArgoCD` seems even more sensible.
> 
> Maybe I'm missing something.
> 
> > Lastly, I prepended the the argo application CR name with the env name since applications would otherwise be likely to overwrite each other.
> 
> I wonder if applications will conflict if they're placed to separate namespaces. I don't really like this change, because it produces more customized internal logic, but let's have it as we both do that at the moment (we in South Pole actually don't need to have this, but my colleague said that he likes this more :-) ).

I guess it is a matter of taste, but let's settle on a default that works out of the box. Currently, looks like you require extra manual config to ensure Argo Application CRs are placed in separate namespaces.

Zebradil commented:
> Shouldn't the user be able to add a schema of sorts, in case he wishes to add his own values, e.g. generic settings that affect multiple applications? We use that already in one of our rollout repos.

The user was able and will be able to modify the schema via `env-data.ytt.yaml` files. I only want to remove `data-schema.ytt.yaml` from the user's sight and manage it by myks.

> I have changed the logic to include an entire schema file (the one from assets) and made if flexible to be overwritten with a schema provided by the user. Please have a look. If you like you can make further changes on this PR so get this feature completed.

Thank you! I'll directly change a couple of tiny bits.

> I guess it is a matter of taste, but let's settle on a default that works out of the box. Currently, looks like you require extra manual config to ensure Argo Application CRs are placed in separate namespaces.

🤝
